### ISSUE TITLE
Consider using the host `beta` toolchain if it already exists at the right version

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#76419]
Currently, if you have `beta` installed via rustup, x.py will download it again when building for the first time. It would be great to avoid this needless download. It would also have the benefit that `x.py clean` wouldn't require redownloading things from the internet.

Possible implementation: try both `rustc --version` and `rustc +beta --version`; use the first that is the right version and download beta otherwise. The version should be an exact match (`beta 1.xx.0`, not just `beta` or `1.xx.0`).

**Comment by lzutao** [comment]
Should nightly rustc newer than beta be considered too ?

**Comment by jyn514** [comment]
I would prefer not because that's a silent change in behavior and you might run into bugs. If you have `rustc = "..."` in config.toml, though, it definitely makes sense to use that version if it exists.

**Comment by cuviper** [comment]
I'm skeptical that this should be done automatically, but you can do it manually:

    ./configure --local-rust-root=$(rustc +beta --print=sysroot)

**Comment by jyn514** [comment]
It seems a waste to redownload the toolchain if it already exists, though. Glad to know it's so easy to implement!

**Comment by jyn514** [comment]
@matklad mentioned in https://matklad.github.io/2020/09/12/rust-in-2021.html that all of this - pinning the version _and_ downloading automatically - could be done by a `rust-toolchain` file in the root of the repository. That would make it easy to opt-out too: just delete the file.

**Comment by Mark-Simulacrum** [comment]
"Just delete the file" means a permanent problem for your git status, which isn't great - but do we know why people would want to opt-out? I think it's a pretty good solution to replace (most of) stage0.txt.

**Comment by jyn514** [comment]
I'm not sure, @cuviper mentioned it in https://github.com/rust-lang/rust/issues/76419#issuecomment-689835847.

**Comment by jyn514** [comment]
One issue with `rust-toolchain` is I don't think it allows having separate toolchains for different components at the same time. `stage0.txt` uses a beta compiler but a nightly rustfmt - @kinnison, @rbtcollins, @workingjubilee is there any plan to support that? Would you take a PR implementing it?

**Comment by matklad** [comment]
Another issue is that it doesn't allow to pin *specific* beta.

**Comment by Mark-Simulacrum** [comment]
I wouldn't worry about rustfmt version being different, we can download that out of band. It's not really critical IMO. (I would rather push through a magic bootstrap env variable to rustfmt and use beta).

I think rust-toolchain already supports a specific beta, e.g. beta-2020-09-29 I would expect works?

**Comment by rbtcollins** [comment]
We also have a toml based overrides file coming in the next release; that documentation doesn't cover it though; I'm not sure where master docs are published at this time.

With regards the rustfmt and compiler toolchain variation; that seems like something we could discuss; though not until the toml structured overrides file has landed. I'd consider it a separate bug and something to design with thought to complications for other projects, support cost, cognitive load etc. But I suspect actually we could make a pretty nice general feature for it, as a number of rust tools rustup installs are indeed pretty separate to the toolchain as a whole.

**Comment by kinnison** [comment]
Another option is to invoke `rustfmt +nightly` if that is plausible.

**Comment by jyn514** [comment]
> Another option is to invoke `rustfmt +nightly` if that is plausible.

The drawback to this is it assumes you're using rustup and gives a hard error otherwise. It should still be possible to bootstrap with a standalone toolchain, it will just be slightly less user friendly.

**Comment by jyn514** [comment]
> One issue with `rust-toolchain` is I don't think it allows having separate toolchains for different components at the same time. `stage0.txt` uses a beta compiler but a nightly rustfmt

We'll be able to side-step this altogether if we download rustfmt on HEAD from CI like I suggested in https://github.com/rust-lang/rust/issues/90672#issuecomment-1200618507 (i.e. `initial_rustfmt` will go away completely).
